oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java

```diff
@@ -0,0 +1,255 @@
+
+        // return original if file too small or split size equals 1
+        if (splitThreshold < IndexerConfiguration.minSplitThreshold() || IndexerConfiguration.splitSize() <= 1) {
+            LOG.info("split is not necessary, skip splitting");
+            return returnOriginalFlatFile();
+        }
+
+        Set<String>splitNodeTypesName = getSplitNodeTypeNames();
+        LOG.info("unsafe split types: {}", splitNodeTypesName);
+        if (splitNodeTypesName.contains(NT_BASE)) {
+            LOG.info("Skipping split because split node types set contains {}", NT_BASE);
+            return returnOriginalFlatFile();
+        }
+
+        try (BufferedReader reader = createReader(flatFile, algorithm)) {
+            long readPos = 0;
+            int outFileIndex = 1;
+            File currentFile = new File(workDir, "split-" + outFileIndex + "-" + getSortedStoreFileName(algorithm));
+            BufferedWriter writer = createWriter(currentFile, algorithm);
+            splitFlatFiles.add(currentFile);
+
+            String line;
+            int lineCount = 0;
+            Stack<String> nodeTypeNameStack = new Stack<>();
+            while ((line = reader.readLine()) != null) {
+                updateNodeTypeStack(nodeTypeNameStack, line);
+                boolean shouldSplit = (readPos > splitThreshold);
+                if (shouldSplit && canSplit(splitNodeTypesName, nodeTypeNameStack)) {
+                    writer.close();
+                    LOG.info("created split flat file {} with size {}", currentFile.getAbsolutePath(), FileUtils.byteCountToDisplaySize(currentFile.length()));
+                    readPos = 0;
+                    outFileIndex++;
+                    currentFile = new File(workDir, "split-" + outFileIndex + "-" + getSortedStoreFileName(algorithm));
+                    writer = createWriter(currentFile, algorithm);
+                    splitFlatFiles.add(currentFile);
+                    LOG.info("split position found at line {}, creating new split file {}", lineCount, currentFile.getAbsolutePath());
+                }
+                writer.append(line);
+                writer.newLine();
+                readPos += line.length() + 1;
+                lineCount++;
+            }
+            writer.close();
+            LOG.info("created split flat file {} with size {}", currentFile.getAbsolutePath(), FileUtils.byteCountToDisplaySize(currentFile.length()));
+
+            LOG.info("split total line count: {}", lineCount);
+        }
+
+        if (deleteOriginal) {
+            LOG.info("removing original flat file {} after splitting into {} files", flatFile.getAbsolutePath(), splitFlatFiles);
+            flatFile.delete();
+        }
+
+        return splitFlatFiles;
+    }
+
+    private void updateNodeTypeStack(Stack<String> parentNodeTypeNames, String line) {
```

Comment:
There are a few util methods in this class like this one - I wonder if it might make sense to make these static to be able to easily write test cases for these ?

Reply:
Possibly we can, but these steps should mostly be covered in the FlatFileSplitterTest. Do you think we are missing tests here.